doc/developer/style-informal.md

```diff
@@ -0,0 +1,66 @@
+# Informal style guide
+
+In addition to the concrete rules, such as for Rust, we also have some more informal style
+guidelines that we try to follow. These are not as strictly enforced as the
+other rules, but you should still try to follow them as much as possible.
+
+## Know your code's scalability limits
+
+When writing code, you should have a good understanding of how it will scale in the future.
+For example, when you build a feature that targets the current scale, you should also think
+about how it will behave when the data size is 10x, 100x, or even 1000x larger.
+What are the bottlenecks? Will it still be performant? Will it still be maintainable?
+
+This doesn't mean that the code you write today has to be perfect for all future scales, but
+it should be easy to change and adapt as the scale changes. Documentation should point out
+any assumptions or limitations that the code has regarding scale. Leave breadcrumbs for your
+future self and others who will work on the code later.
+
+A common way to think about scalability is to keep time and space complexity in mind.
+For example, if you are writing a function that takes 50 microseconds per item, it is
+fine to write a O(n^2) algorithm for very small n, but it will quickly become a problem.
+The following table demonstrates how different time complexities scale with increasing input sizes:
+
+| Input Size (n) | O(1)     | O(log n) | O(n)  | O(n log n) | O(n^2)     |
+|----------------|----------|----------|-------|------------|------------|
+| 10             | constant | 166us    | 500us | 1.66ms     | 5ms        |
+| 100            | constant | 333us    | 5ms   | 33ms       | 500ms      |
+| 1,000          | constant | 500us    | 50ms  | ~500ms     | 50s        |
+| 10,000         | constant | 665us    | 500ms | ~6.7s      | 84 minutes |
+| 100,000        | constant | 831us    | 5s    | ~83s       | 139h       |
+| 1,000,000      | constant | 1ms      | 50s   | ~1000s     | 1.6 years  |
+
+## Prefer simplicity to complexity
```

Comment:
would love to include some of Frank's heuristics here as well! https://www.notion.so/materialize/Technical-Leadership-19e13f48d37b80a9b170db4170accd29?source=copy_link#19e13f48d37b806c8778d1a1d18655cb